src/librustc_yk_sections/mir_cfg.rs

```diff
@@ -18,20 +16,95 @@ use rustc::ty::TyCtxt;
 
 use rustc::hir::def_id::DefId;
 use rustc::mir::{
-    Mir, TerminatorKind, Operand, Constant, StatementKind, BasicBlock, BasicBlockData, Terminator
+    Mir, TerminatorKind, Operand, Constant, StatementKind, BasicBlock, BasicBlockData, Terminator,
+    Place, Rvalue, Statement, Local, PlaceBase
 };
 use rustc::ty::{TyS, TyKind, Const, LazyConst};
 use rustc::util::nodemap::DefIdSet;
 use std::path::PathBuf;
 use std::fs::File;
+use std::convert::TryFrom;
 use rustc_yk_link::YkExtraLinkObject;
 use std::fs;
 use std::error::Error;
+use std::cell::{Cell, RefCell};
+use rustc_data_structures::indexed_vec::{IndexVec, Idx};
+use rustc_data_structures::graph::dominators::Dominators;
 use ykpack;
 
 const SECTION_NAME: &'static str = ".yk_tir";
 const TMP_EXT: &'static str = ".yk_tir.tmp";
 
+type TirVarIndex = u32;
+
+/// A conversion context holds the state needed to perform the conversion to (pre-SSA) TIR.
+struct ConvCx<'a, 'tcx, 'gcx> {
+    /// The compiler's god struct. Needed for queries etc.
+    tcx: &'a TyCtxt<'a, 'tcx, 'gcx>,
+    /// The definition sites of TIR variable in terms of basic blocks.
+    def_sites: RefCell<Vec<Vec<BasicBlock>>>,
+    /// The next TIR variable index to be issued.
+    next_tir_var: Cell<TirVarIndex>,
+    /// A mapping from MIR variables to TIR variables.
+    var_map: RefCell<IndexVec<Local, Option<TirVarIndex>>>,
+}
+
+impl<'a, 'tcx, 'gcx> ConvCx<'a, 'tcx, 'gcx> {
+    fn new(tcx: &'a TyCtxt<'a, 'tcx, 'gcx>, _mir: &Mir<'tcx>) -> Self {
+        let var_map = IndexVec::new();
+
+        Self {
+            tcx,
+            def_sites: RefCell::new(Vec::new()),
+            next_tir_var: Cell::new(0),
+            var_map: RefCell::new(var_map),
+        }
+    }
+
+    /// Issues a fresh TIR variable index.
+    fn next_tir_var(&self) -> TirVarIndex {
+        let var_idx = self.next_tir_var.get();
+        self.next_tir_var.set(var_idx + 1);
+        var_idx
+    }
+
+    /// Get the TIR variable corresponding with the MIR variable, creating a fresh one if needed.
+    fn get_tir_var(&self, local: Local) -> TirVarIndex {
+        let local_u32 = local.as_u32();
+        let mut var_map = self.var_map.borrow_mut();
+
+        // Resize the backing Vec if necessary.
+        if var_map.len() <= usize::try_from(local_u32).unwrap() {
```

Comment:
I would prefer to do this as:

```rust
debug_assert!(usize::max_value >= u32::max_value());
if var_map.len <= usize::from(local.as_u32()) { ... }
```

because the `try_from` is never really necessary (we're never going to run on a 16 bit machine) so we might as well avoid the performance hit.

Reply:
Ah, I was meant to ask about try_from in the PR description.

We have to convert in both directions. From u32 to usize and back.

The u32 to usize direction can be solved as you say. What about the inverse?

Reply:
In general usize to u32 is clearly unsafe on a 64 bit machine. *But* if we know that we only ever wrote out 32 bit numbers, the inverse is safe. It's fine to do something like this:

```rust
// Since we only ever issue numbers which fit into 32 bits, this conversion
// is guaranteed to be safe.
let x: usize = ...;
let y: u32: x as u32;
```

But each case must be clearly documented.

Reply:
OK, seems fine by me.

[I think Rust really makes this harder than it needs to be :( ]

Reply:
Don't start me on my Rust-integer-conversions-aren't-great rant!

Reply:
I wonder if we actually want to check the size of the integer in bytes, not the maximum values:

```
77 |         debug_assert!(usize::max_value() >= u32::max_value());
   |                                             ^^^^^^^^^^^^^^^^ expected usize, found u32
```

If we don't know which is larger out of usize and u32, we can't insert a correct cast on either the left or right hand side of the assertion, right?

So should we use?:
```
debug_assert!(size_of(usize) >= size_of(u32));
```

I think that's robust as long as we are careful not to mix signedness.

Reply:
In practise these are clearly equivalent *but* in theory the size of a thing doesn't guarantee the range of values it can contain (Rust makes no guarantees about bit representations IIRC), so `max_value` is perhaps marginally safer in an angels-on-the-head-of-a-pin sense.

Reply:
So I think you are suggesting that we assume usize is wider than u32 and we do:
```
debug_assert!(usize::max_value() >= usize::from(u32::max_value()));
```

But if we are assuming exactly what we are asserting, then why have the assertion at all?

Maybe I'm missing something?

Worded differently, how would you fix the above compilation error in a meaningful way?

Reply:
The point of the assertion is that if, somehow, this gets compiled on a platform where the assertion isn't true, it'll leave behind a panic. Thus the assertion implicitly guarantees that our assumption doesn't silently cause problems.

Reply:
Yes, I understand that :)

I think I'm failing to communicate my point. Let me try again.

You suggested an assertion like this:
```
debug_assert!(usize::max_value() >= u32::max_value());
```

This can't work because the max value on the left hand side of the `>=` evaluates to a `usize` and the max value on the right is a `u32`. So we cannot directly compare them.

We'd need a safe cast on either the left or right, but I can't see how if you don't know which of a `usize` or a `u32` is bigger.

We can assume a `usize` can express all of the values of a `u32` and do:
```
debug_assert!(usize::max_value() >= usize::from(u32::max_value()));
```

But if we assume a `usize` can express all of the values of a `u32`, then the assertion is a tautology from the start.

Do you follow?

Reply:
Ah, I'm with you! Yes, you're right. In fact, my past self knew this too ;)

```
$ cd grmtools
$ rg size_of|grep assert
cfgrammar/src/lib/idxnewtype.rs:                debug_assert!(size_of::<usize>() >= size_of::<T>());
cfgrammar/src/lib/idxnewtype.rs:                debug_assert!(size_of::<u32>() >= size_of::<T>());
lrpar/src/lib/lex.rs:        debug_assert!(size_of::<usize>() >= size_of::<u32>());
lrpar/src/lib/lex.rs:            debug_assert!(size_of::<usize>() >= size_of::<u32>());
lrtable/src/lib/mod.rs:        debug_assert!(size_of::<usize>() >= size_of::<StIdxStorageT>());
```

Reply:
Cool!

Reply:
By the way, in cases like:

```
debug_assert!(size_of::<usize>() >= size_of::<T>());
```

Might be worth double checking `T` must be unsigned?

Reply:
Yes, that is enforced through the `num_traits` crate.

Reply:
Cool, one more thing, you suggested:
```
if var_map.len <= usize::from(local.as_u32()) { ... }
```

But I think it'll need to be as `as` conversion, due to:
```
79 |         if var_map.len() <= usize::from(local_u32).unwrap() {
   |                             ^^^^^^^^^^^ the trait `std::convert::From<u32>` is not implemented for `usize`
```

Is that OK?

Reply:
The assertion guarantees that the `as` is safe, right? But, in general, we should add a comment at each `as` site documenting why it's safe (e.g. "this is guaranteed by the execution in `f`, which must have been called before this function").

Reply:
Agreed on both counts.